File: app/src/main/java/xyz/arnau/muvicat/cache/db/MuvicatDatabase.java
package xyz.arnau.muvicat.cache.db;

import android.arch.persistence.room.Database;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.RoomDatabase;
import android.arch.persistence.room.TypeConverters;
import android.content.Context;

import xyz.arnau.muvicat.AppExecutors;
import xyz.arnau.muvicat.cache.dao.CinemaDao;
import xyz.arnau.muvicat.cache.dao.MovieDao;
import xyz.arnau.muvicat.cache.dao.PostalCodeDao;
import xyz.arnau.muvicat.cache.dao.ShowingDao;
import xyz.arnau.muvicat.cache.db.migrations.DbMigration1to2;
import xyz.arnau.muvicat.cache.db.migrations.DbMigration2to3;
import xyz.arnau.muvicat.cache.model.CinemaEntity;
import xyz.arnau.muvicat.cache.model.MovieEntity;
import xyz.arnau.muvicat.cache.model.PostalCodeEntity;
import xyz.arnau.muvicat.cache.model.ShowingEntity;
import xyz.arnau.muvicat.cache.utils.PostalCodeCsvReader;

@Database(
        entities = {MovieEntity.class, CinemaEntity.class, PostalCodeEntity.class, ShowingEntity.class},
        version = 4
)
@TypeConverters(DateTypeConverter.class)
public abstract class MuvicatDatabase extends RoomDatabase {
    private static final String MUVICAT_DB = "muvicat-db";
    private static MuvicatDatabase sInstance = null;

    public abstract MovieDao movieDao();
    public abstract CinemaDao cinemaDao();
    public abstract PostalCodeDao postalCodeDao();
    public abstract ShowingDao showingDao();

    @SuppressWarnings("SynchronizedMethod")
    public static synchronized MuvicatDatabase getInstance(Context context, AppExecutors appExecutors) {
        if (sInstance == null) {
            sInstance = buildDatabase(context, appExecutors);
        }
        return sInstance;
    }

    private static MuvicatDatabase buildDatabase(Context context, AppExecutors appExecutors) {
        return Room.databaseBuilder(context, MuvicatDatabase.class, MUVICAT_DB)
                .addMigrations(DbMigration1to2.INSTANCE, DbMigration2to3.INSTANCE)
                .addCallback(new PostalCodesDbCallback(context, appExecutors, new PostalCodeCsvReader()))
                .build();
    }
}

File: app/src/main/java/xyz/arnau/muvicat/cache/db/MuvicatDatabase.kt
package xyz.arnau.muvicat.cache.db

import android.arch.persistence.room.Database
import android.arch.persistence.room.Room
import android.arch.persistence.room.RoomDatabase
import android.arch.persistence.room.TypeConverters
import android.content.Context
import xyz.arnau.muvicat.AppExecutors
import xyz.arnau.muvicat.cache.dao.CinemaDao
import xyz.arnau.muvicat.cache.dao.MovieDao
import xyz.arnau.muvicat.cache.dao.PostalCodeDao
import xyz.arnau.muvicat.cache.dao.ShowingDao
import xyz.arnau.muvicat.cache.db.migrations.*
import xyz.arnau.muvicat.cache.utils.PostalCodeCsvReader
import xyz.arnau.muvicat.data.model.Cinema
import xyz.arnau.muvicat.data.model.Movie
import xyz.arnau.muvicat.data.model.PostalCode
import xyz.arnau.muvicat.data.model.Showing

@Database(entities = [Movie::class, Cinema::class, PostalCode::class, Showing::class], version = 4)
@TypeConverters(DateTypeConverter::class)
abstract class MuvicatDatabase : RoomDatabase() {
    abstract fun movieDao(): MovieDao
    abstract fun cinemaDao(): CinemaDao
    abstract fun postalCodeDao(): PostalCodeDao
    abstract fun showingDao(): ShowingDao

    companion object {
        private var INSTANCE: MuvicatDatabase? = null

        @Synchronized
        fun getInstance(context: Context, appExecutors: AppExecutors): MuvicatDatabase {
            if (INSTANCE == null) {
                INSTANCE = buildDatabase(context, appExecutors)
            }
            return INSTANCE!!
        }

        private fun buildDatabase(context: Context, appExecutors: AppExecutors): MuvicatDatabase {
            return Room.databaseBuilder(context, MuvicatDatabase::class.java, "muvicat-db")
                .addMigrations(
                    DbMigration1to2,
                    DbMigration2to3,
                    DbMigration3to4
                )
                .addCallback(PostalCodesDbCallback(context, appExecutors, PostalCodeCsvReader()))
                .build()
        }
    }
}